src/cheri_insts.sail

```diff
@@ -1308,26 +1316,34 @@ function handle_store_cond_cap_via_cap(cs2, cs, cap_val, vaddrBits) = {
   } else {
     match translateAddr(vaddrBits, Write(if cs2_val.tag then Cap else Data)) {
       TR_Failure(e, _) => { handle_mem_exception(vaddrBits, e); RETIRE_FAIL },
-      TR_Address(addr, _) => {
-        let eares : MemoryOpResult(unit) = mem_write_ea_cap(addr, aq, rl, false);
-        match (eares) {
-          MemException(e) => { handle_mem_exception(addr, e); RETIRE_FAIL },
-          MemValue(_) => {
-            let res : MemoryOpResult(bool) = mem_write_cap(addr, cs2_val, aq, rl, false);
-            match (res) {
-              MemValue(true)  => {
-                 writeCapReg(cs2, int_to_cap(EXTZ(0b0)));
-                 cancel_reservation();
-                 RETIRE_SUCCESS
-              },
-              MemValue(false) => {
-                 writeCapReg(cs2, int_to_cap(EXTZ(0b1)));
-                 cancel_reservation();
-                 RETIRE_SUCCESS
-              },
-              MemException(e) => {
-                handle_mem_exception(addr, e);
-                RETIRE_FAIL
+      TR_Address(addr, caveats) => {
+        match (caveats.ptw_sc_mod, cs2_val.tag) {
+          (PTW_SCM_TRAP, true) => {
+            handle_cheri_cap_exception(CapEx_TLBNoStoreCap, cs);
+            RETIRE_FAIL
+          },
+          _ => {
+            let eares : MemoryOpResult(unit) = mem_write_ea_cap(addr, aq, rl, false);
+            match (eares) {
+              MemException(e) => { handle_mem_exception(addr, e); RETIRE_FAIL },
+              MemValue(_) => {
+                let res : MemoryOpResult(bool) = mem_write_cap(addr, cs2_val, aq, rl, false);
+                match (res) {
+                  MemValue(true)  => {
+                     writeCapReg(cs2, int_to_cap(EXTZ(0b0)));
+                     cancel_reservation();
+                     RETIRE_SUCCESS
+                  },
+                  MemValue(false) => {
+                     writeCapReg(cs2, int_to_cap(EXTZ(0b1)));
+                     cancel_reservation();
+                     RETIRE_SUCCESS
+                  },
```

Comment:
Surely whilst we're here we can make this code less crazy as:
```suggestion
                  MemValue(success)  => {
                     writeCapReg(cs2, int_to_cap(EXTZ(bool_to_bits(!success))));
                     cancel_reservation();
                     RETIRE_SUCCESS
                  },
```
and at some point maybe do the same to sail-riscv's A extension implementation (though `X(rd) = EXTZ(0b0);` is at least more succinct than `writeCapReg(cs2, int_to_cap(EXTZ(0b0)));`...).

Reply:
That may be, but I think in light of your comment above, we are no longer here, so I'll leave this to a later cleanup.